### PR TITLE
Selftest rework, fan selftest + fanctl fix

### PR DESCRIPTION
**Pull request by XPila**
BFW-1278

### CODE REVIEW

File: src/common/selftest/selftest_heater.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// selftest_heater.h
+#pragma once
+
+#include <inttypes.h>
+#include "selftest_MINI.h"
+
+typedef struct _selftest_heater_config_t {
```
Comment (RadekVana) [inline src/common/selftest/selftest_heater.h:7]: why C struct?

File: src/common/selftest/selftest_axis.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    };
+
+public:
+    CSelftestPart_Axis(const selftest_axis_config_t *pconfig);
+
+public:
+    bool IsInProgress() const;
+
+public:
+    bool Start();
```
Comment (RadekVana) [inline src/common/selftest/selftest_axis.h:35]: parent has this method virtual, so this one is virtual too. Use virtual and override so it is obvious
virtual  bool Start() override;

File: src/common/selftest/selftest_fan.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    };
+
+public:
+    CSelftestPart_Fan(const selftest_fan_config_t *pconfig);
+
+public:
+    bool IsInProgress() const;
+
+public:
+    bool Start();
```
Comment (RadekVana) [inline src/common/selftest/selftest_fan.h:36]: parent has this method virtual, so this one is virtual too. Use virtual and override so it is obvious
virtual  bool Start() override;

File: src/common/fanctl.h [hunk 4]
```diff
@@ -134,15 +135,16 @@ class CFanCtl {
...
+    uint32_t measure();                  // measure tacho delay [us], blocking
 private:
-    uint16_t m_MinRPM;  // minimum rpm value (set in constructor)
-    uint16_t m_MaxRPM;  // maximum rpm value (set in constructor)
-    FanState m_State;   // fan control state
-    uint8_t m_PWMValue; // current pwm value
-    uint8_t m_Edges;    // edge counter - used for starting and measurement
-    uint8_t m_Ticks;    // tick counter - used for starting and measurement
-
+    uint16_t m_MinRPM;   // minimum rpm value (set in constructor)
```
[the remark this answers is not in the document]
Comment (XPila) [reply src/common/fanctl.h]: fixed

File: src/common/fanctl.cpp [hunk 5]
```diff
@@ -183,16 +184,26 @@ void CFanCtl::tick() {
...
+        if (m_Ticks > 1000) {
+            m_Result = 0;
+            m_State = blanking;
+        } else {
             m_pwm.set_PWM(m_pwm.get_max_PWM());
             if (edge)
                 m_Edges++;
-            if (m_Edges >= 2)
-                m_State = running;
+            if (m_Edges >= 3) {
```
[the remark this answers is not in the document]
Comment (XPila) [reply src/common/fanctl.cpp]: fixed also other numbers

File: src/common/selftest/selftest_heater.h (hunk 1 above)
Comment (XPila) [reply src/common/selftest/selftest_heater.h:7]: why not? i need simple define entire structure as constant (seperate the members from the class). the structure is used in constructor and in static method estimate (to calculate total time of the test).

File: src/common/selftest/selftest_axis.h (hunk 2 above)
Comment (XPila) [reply src/common/selftest/selftest_axis.h]: fixed

File: src/common/selftest/selftest_fan.h (hunk 3 above)
Comment (XPila) [reply src/common/selftest/selftest_fan.h]: fixed

**Review by RadekVana — APPROVED** [review]

**Pull request merged**